### PR TITLE
Invitation letter controller

### PR DESCRIPTION
**Pull request by soginga**
#329

### CODE REVIEW

File: api/app/invitationletter/api.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+        global invitation_template
+
+        if offer.accommodation_award and offer.travel_award:
+            invitation_template = db.session.query(InvitationTemplate).filter(
+                InvitationTemplate.event_id == offer.event_id).filter(
+                InvitationTemplate.send_for_both_travel_accommodation).first()
+        elif offer.travel_award:
+            invitation_template = db.session.query(InvitationTemplate).filter(
+                InvitationTemplate.event_id == offer.event_id).filter(
+                InvitationTemplate.send_for_both_travel_accommodation).first()
```
[the remark this answers is not in the document]
Comment (soginga) [reply api/app/invitationletter/api.py]: thanks! overlooked those.

File: api/app/invitationletter/tests.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        data = json.loads(response.data)
+        header = {'Authorization': data['token']}
+        return header
+
+    def test_create_create_invitation_letter(self):
+        self.seed_static_data()
+        response = self.app.post(
+                '/api/v1/invitation-letter', data=INVITATION_LETTER, headers=self.headers)
+        data = json.loads(response.data)
+        LOGGER.debug(
```
Comment (jaderabbit) [inline api/app/invitationletter/tests.py]: can you assert the fields are saved correctly?

Comment (soginga) [reply api/app/invitationletter/tests.py]: That's not possible given i'm supposed to return only invitation_letter_request_id from the post request #329 or should i modified what get's returned from the api to InvitationLetterRequest fields?

Comment (jaderabbit) [reply api/app/invitationletter/tests.py]: @soginga You should go retrieve it from the database and check it's updated. Its a test. This is how one tests :P

File: api/app/invitationletter/api.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+            from_date=from_date,
+        )
+
+        offer = db.session.query(Offer).filter(
+            Offer.user_id == user_id).first()
+
+        if not offer:
+            return errors.OFFER_NOT_FOUND
+
+        global invitation_template
```
Comment (avishkar58) [inline api/app/invitationletter/api.py]: Why is this global?

File: api/app/invitationletter/api.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+        if offer.accommodation_award and offer.travel_award:
```
Comment (avishkar58) [inline api/app/invitationletter/api.py]: Please also check accepted_accommodation_award and accepted_travel_award, so the condition should be something like: 
```if offer.accommodation_award and offer.accepted_accommodation_award and offer.travel_award and offer.accepted_travel_award:```
Same applies for the below two

File: api/app/invitationletter/api.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+                InvitationTemplate.send_for_both_travel_accommodation).first()
+        elif offer.travel_award:
+            invitation_template = db.session.query(InvitationTemplate).filter(
+                InvitationTemplate.event_id == offer.event_id).filter(
+                InvitationTemplate.send_for_travel_award_only).first()
+        elif offer.accommodation_award:
+            invitation_template = db.session.query(InvitationTemplate).filter(
+                InvitationTemplate.event_id == offer.event_id).filter(
+                InvitationTemplate.send_for_accommodation_award_only).first()
+
```
Comment (avishkar58) [inline api/app/invitationletter/api.py]: What about the base case for people who have no travel or accommodation award?

File: api/app/invitationletter/api.py (hunk 3 above)
Comment (jaderabbit) [reply api/app/invitationletter/api.py]: OMG YES WHY YOU DO DIS? :scream:

**Comment by jaderabbit** [comment]
@soginga We've updated the issue. 

We ask during registration if they're bringing a poster. 
The template will have %BRING_POSTER%

- *If*, in their registration, they said they're bringing a poster, then we should populate %BRING POSTER% with the following (temporary, to be updated) text:
```
They will be presenting an academic poster of their research at the event, entitled: <Poster Title>
```
These details you can get from registration 

- *Else*, populate %BRING_POSTER% with an empty string

File: api/app/invitationletter/api.py (hunk 3 above)
Comment (soginga) [reply api/app/invitationletter/api.py]: mm🤔 yea it's kinda dumb I'm not sure... I was getting the warning that `"Local variable 'invitation_template' might be referenced before the assignment"` looked it up, they suggested that on stackoverflow

**Comment by avishkar58** [comment]
Better than global would be to use `invitation_template = None`. Better
still would be to think about what the warning is telling you - you might
be using invitation_template without having assigned it a value - ask
yourself whether this is a valid state for your program. In this case, it
isn't, because when it comes time to use invitation_template to generate
the letter, you expect it to have been populated with one of the templates.
So the best solution would be to remove that line entirely and instead
handle the ELSE case to set invitation_template for people who have no
awards :)

File: api/app/invitationletter/api.py (hunk 3 above)
Comment (soginga) [reply api/app/invitationletter/api.py]: I was getting the `Local variable 'invitation_template' might be referenced before assignment`

File: api/app/invitationletter/api.py [hunk 6]
```diff
@@ -57,21 +57,27 @@ def post(self):
...
                 InvitationTemplate.send_for_travel_award_only).first()
-        elif offer.accommodation_award:
+
+        elif offer.accommodation_award and offer.accepted_accommodation_award:
             invitation_template = db.session.query(InvitationTemplate).filter(
                 InvitationTemplate.event_id == offer.event_id).filter(
                 InvitationTemplate.send_for_accommodation_award_only).first()
 
+        elif (not offer.accommodation_award) and (not offer.travel_award):
+            return errors.NO_ACCOMMODATION_TRAVEL_AWARD
```
Comment (jaderabbit) [inline api/app/invitationletter/api.py]: It's not an error case. They also get a invitation letter even if they don't get an award. There is a default invitation letter

File: api/app/utils/errors.py [hunk 7]
```diff
@@ -65,3 +65,5 @@
     {'message': 'Offer has not been accepted'}, 409)
 APPLICATIONS_CLOSED = (
     {'message': 'Applications are now closed'}, 403)
+NO_ACCOMMODATION_TRAVEL_AWARD = (
+    {'message': 'No template found because your offer has neither accommodation nor travel award allocation'}, 404)
```
Comment (jaderabbit) [inline api/app/utils/errors.py]: There is a template if you have neither.
You still get an invitation letter even if they're not sponsoring travel or accomodation :P

File: api/app/invitationletter/api.py (hunk 6 above)
Comment (soginga) [reply api/app/invitationletter/api.py]: so what's the condition to retrieve that template? 

> `....filter( not InvitationTemplate.send_for_both_travel_accommodation).filter( not InvitationTemplate.send_for_travel_award_only).filter( not InvitationTemplate.send_for_accommodation_award_only)` ?

Comment (jaderabbit) [reply api/app/invitationletter/api.py]: All things false :)

File: api/app/invitationletter/api.py [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+                               passport_issued_by=passport_issued_by,
+                               invitation_letter_sent_at=invitation_letter_request.invitation_letter_sent_at,
+                               to_date=to_date,
+                               from_date=from_date,
+                               country_of_residence=country_of_residence,
+                               nationality=nationality,
+                               date_of_birth=date_of_birth,
+                               email=user.email)
+
+            if is_sent:
```
Comment (jaderabbit) [inline api/app/invitationletter/api.py]: if it's not sent, then we also need a failcase. 
"Failed to send Invitation Email"

**Pull request merged**